### PR TITLE
Emit per-parameter `INFO` when `inferSpec` drops the signature

### LINKED ISSUE #491: Add full-⊤ (dtype-⊤ and shape-⊤) regression fixture
## Background

Follow-up to #486. The lattice has two orthogonal axes per the wala/ML `com.ibm.wala.cast.python.ml/CONTRIBUTING.md` contract:

- Shape: `null` dims is shape-⊤.
- Dtype: `EnumSet.of(DType.UNKNOWN)` is dtype-⊤.

The regression test added by #486's fix (`testInferredTensorTypesUnknownShapeTop`) exercises shape-⊤ (`tf.keras.Input(shape=json.loads(...))` produces `TensorType(FLOAT32, null)`—dtype is `tf.keras.Input`'s default `float32`). The dtype-⊤ axis is not exercised.

## What's Needed

A fixture where Ariadne genuinely cannot determine the dtype, producing a `TensorType` with `getDType() == DType.UNKNOWN`. Combined with shape-⊤ (`null` dims), the full-⊤ marker `(DType.UNKNOWN, null)` would be pinned.

## Candidate APIs to Probe

- `tf.py_function(...)` with no `Tout` spec.
- `tf.cast(t, dtype=<dynamic>)` where the target dtype is opaque.
- A parameter reached only via `**kwargs` from an unanalyzable call site.
- Composition with `tf.numpy_function` or similar Python-bridge APIs.

Each candidate needs Ariadne probing to confirm it produces dtype-⊤ rather than a hardcoded default.

## Scope

Read-only investigation first (probe each candidate, log what `Parameter.getTensorTypes()` returns), then add the fixture and assertion. If no candidate reliably produces dtype-⊤, file an Ariadne tracker for an API that should but doesn't (or accept that the dtype-⊤ axis is rarely exercised in practice).

## Cross-Refs

- #486 (lattice violation, original).
- `com.ibm.wala.cast.python.ml/CONTRIBUTING.md` § `Dtypes—getDefaultDTypes` for the contract.

### PR DESCRIPTION
**Pull request by khatchad**
## Summary

Closes #510, closes #494, closes #491. `inferInputSignature` already emits an `INPUT_SIGNATURE_INFERENCE` INFO for its category (a) (non-tensor) and category (b) (no Phase-2 evidence) drops, but the `inferSpec`-side drop collapsed the signature silently. This adds a reason-aware per-parameter INFO at that drop site, with fixtures covering both reachable drop reasons (heterogeneous dtype and dtype-⊤).

## Re-Scope Note

#510 was written against the pre-#480 `inferSpec` and lists four drop branches. Against the current `inferSpec`, only two of them still drop:

- Heterogeneous dtype (`|D| ≠ 1`) — reachable and tested.
- Dtype-⊤ (a single agreed `UNKNOWN`) — reachable via `tf.constant(np_array)` (wala/ML#539) and tested.

The other branches #510 listed no longer drop: #480's per-context reduction emits a coarse `TensorType(dtype, null)` (shape-⊤) or a `SymbolicDim` wildcard (per-dim disagreement) instead, and multi-context is reduced rather than dropped. The INFO distinguishes the two real reasons in plain, wizard-facing prose (the message shows on the refactoring wizard).

## Test

Two fixtures, both asserting an INFO-severity `INPUT_SIGNATURE_INFERENCE` status citing the parameter and the drop reason:

- `testHasLikelyTensorParameter15` (the existing `double` fixture, INT32/FLOAT32/STRING) — the heterogeneous-dtype drop.
- `testInputSignatureDtypeTop` (new, `consume(tf.constant(np.ones(..., dtype=...)))`) — the dtype-⊤ drop, pinning the current wala/ML#539 behavior; it also covers #494's dtype-⊤ singleton and #491's full-⊤ marker, and inverts when wala/ML#539 lands.

Full suite green (510, 0 failures).

## Cross-Refs

- #510 — this PR.
- #480 — introduced the per-context reduction that narrowed `inferSpec`'s drop set.
- #494 — dtype-⊤ branch coverage; this PR's `testInputSignatureDtypeTop` covers it.
- #491 — full-⊤ marker; covered by the same fixture.
- wala/ML#539 — the `tf.constant(np_array)` dtype-loss that makes the dtype-⊤ branch reachable (and that the pinning test tracks).
- #508 — the category (a)/(b) INFO design this extends.

### CODE REVIEW

File: edu.cuny.hunter.hybridize.core/src/edu/cuny/hunter/hybridize/core/analysis/Function.java [hunk 1]
```diff
@@ -1556,8 +1556,19 @@ public Optional<InputSignature> inferInputSignature() {
...
+				 * heterogeneous dtype (|D| ≠ 1) or dtype-⊤ (a single agreed `UNKNOWN`). Shape-⊤ and symbolic-dim no longer drop here—it
+				 * emits a coarse `TensorType(dtype, null)` or a `SymbolicDim` wildcard instead. Emit a per-parameter INFO naming the
+				 * reason; see https://github.com/ponder-lab/Hybridize-Functions-Refactoring/issues/510.
+				 */
+				if (contexts.stream().map(TensorType::getDType).distinct().count() > 1)
+					this.addInfo(INPUT_SIGNATURE_INFERENCE, "Parameter `" + param.getName() + "` of `" + this
+							+ "` receives tensors with conflicting dtypes across call sites, so a single input signature cannot be inferred; it is dropped.");
+				else
+					this.addInfo(INPUT_SIGNATURE_INFERENCE, "Parameter `" + param.getName() + "` of `" + this
+							+ "` receives a tensor whose dtype cannot be determined, so a single input signature cannot be inferred; it is dropped.");
```
[the remark this answers is not in the document]
Comment (khatchad) [reply edu.cuny.hunter.hybridize.core/src/edu/cuny/hunter/hybridize/core/analysis/Function.java:1571]: > it doesn't explicitly identify `tf.UNKNOWN` / dtype-⊤ nor cite the tracked coverage issue (#494)

Intentional. This INFO surfaces on the refactoring wizard, so it is written as plain end-user prose: the lattice jargon (`tf.UNKNOWN`, dtype-⊤) and the issue number were deliberately dropped because they read as odd to someone running the refactoring. The technical terms and the #494 reference live in the code comment and the PR description (developer-facing), not in the user-facing string.

> the current `distinct().count()` check only distinguishes ">1" vs "otherwise", which makes it hard to keep the message aligned with `inferSpec`'s actual drop reasons as that logic evolves

Fair point. The caller re-derives the reason from `contexts` rather than receiving it from `inferSpec`, so the two could drift. The robust fix is #483 (graduate `inferInputSignature` to a typed `Result<InputSignature, AbsenceReason>` carrying the drop reason), which is its own card; this PR takes the lightweight caller-side approach pending that. Today there are only two reachable drop reasons (heterogeneous dtype vs dtype-⊤), which the `>1` split distinguishes correctly.
